### ISSUE TITLE
"Cannot revive if not dead"

### ISSUE THREAD

**Issue by lemoness** [HabitRPG/habitica#9478]
We've gotten several reports that when users die and attempt to revive themselves via the popup modal by clicking "Refill Health and Try Again", they get the error message "Cannot revive if not dead" and have to refresh before they can restore their health.

**Comment by TheHollidayInn** [comment]
I've seen these reports in the bug guild which seemed to be related to VersionErrors on the notifications array. I'm slowly making improvements, but if we can log reports here, that would help confirm the VersionError is the only issue.

**Comment by Alys** [comment]
Celestine, Rogue Lady of the Moon (9e9c1391-f10c-4e07-9e72-c582daabb3a5) reports this:
"Got that can't revive if not dead thing again (the refill health modal appeared three times)."

According to amplitude, the death event was at January 10, 2018 2:35:27 pm UTC. That's 2018-01-11 00:35:27 in my timezone, which is what's used in the Loggly errors below.

Loggly reports no VersionError. @TheHollidayInn 

In case more Loggly details are helpful, although I doubt they are:

The event in Loggly before the death was an unrelated ""NotFound: Challenge not found or you don't have access" message at 2018-01-10 10:33:34.351

Then Loggly has two "NotAuthorized: Cannot revive if not dead" messages:

2018-01-11 00:35:30.890
{ json: { isHandledError: true, headers: { referer: "https://habitica.com/", x-arr-ssl: "2048|256|C=GB, S=Greater Manchester, L=Salford, O=COMODO CA Limited, CN=COMODO RSA Domain Validation Secure Server CA|OU=Domain Control Validated, OU=EssentialSSL Wildcard, CN=*.habitica.com", content-length: "0", accept-language: "en-US,en;q=0.5", x-forwarded-proto: "https", x-forwarded-port: "443", x-forwarded-for: "......", x-api-user: "9e9c1391-f10c-4e07-9e72-c582daabb3a5", x-original-host: "habitica.com", accept: "application/json, text/plain, */*", x-user-timezoneoffset: "..", x-original-url: "/api/v3/user/revive", host: "habitica.com", x-client: "habitica-web", max-forwards: "10", connection: "Keep-Alive", accept-encoding: "gzip, deflate, br", x-arr-log-id: "0eed6635-4d7d-44fb-94ed-69da53db7a63", user-agent: "..." }, method: "POST", level: "warn", originalUrl: "/api/v3/user/revive", httpCode: 401, message: "NotAuthorized: Cannot revive if not dead at Object.revive (/usr/src/habitrpg/website/common/transpiled-babel/ops/revive.js:46:11) [etc]

2018-01-11 00:35:53.382
{ json: { isHandledError: true, headers: { referer: "https://habitica.com/", x-arr-ssl: "2048|256|C=GB, S=Greater Manchester, L=Salford, O=COMODO CA Limited, CN=COMODO RSA Domain Validation Secure Server CA|OU=Domain Control Validated, OU=EssentialSSL Wildcard, CN=*.habitica.com", content-length: "0", accept-language: "en-US,en;q=0.5", x-forwarded-proto: "https", x-forwarded-port: "443", x-forwarded-for: "....", x-api-user: "9e9c1391-f10c-4e07-9e72-c582daabb3a5", x-original-host: "habitica.com", accept: "application/json, text/plain, */*", x-user-timezoneoffset: "..", x-original-url: "/api/v3/user/revive", host: "habitica.com", x-client: "habitica-web", max-forwards: "10", connection: "Keep-Alive", accept-encoding: "gzip, deflate, br", x-arr-log-id: "78c6637c-5d8f-4720-98f8-6c61b404b084", user-agent: "..." }, method: "POST", level: "warn", originalUrl: "/api/v3/user/revive", httpCode: 401, message: "NotAuthorized: Cannot revive if not dead [etc]


And the next thing in Loggly is another unrelated message several minutes later from a Daily or To-Do being completed when it had already been marked as completed:

2018-01-11 00:46:19.014
{ json: { isHandledError: true, headers: { referer: "https://habitica.com/", x-arr-ssl: "2048|256|C=GB, S=Greater Manchester, L=Salford, O=COMODO CA Limited, CN=COMODO RSA Domain Validation Secure Server CA|OU=Domain Control Validated, OU=EssentialSSL Wildcard, CN=*.habitica.com", content-length: "0", accept-language: "en-US,en;q=0.5", x-forwarded-proto: "https", x-forwarded-port: "443", x-forwarded-for: "...", x-api-user: "9e9c1391-f10c-4e07-9e72-c582daabb3a5", x-original-host: "habitica.com", accept: "application/json, text/plain, */*", x-user-timezoneoffset: "...", x-original-url: "/api/v3/tasks/4359a035-9214-4b9e-bbbe-a03e5316a562/score/up", host: "habitica.com", x-client: "habitica-web", max-forwards: "10", connection: "Keep-Alive", accept-encoding: "gzip, deflate, br", x-arr-log-id: "8f5f0871-eea5-41f8-8cd3-4cd2d80be530", user-agent: "..." }, method: "POST", level: "warn", originalUrl: "/api/v3/tasks/4359a035-9214-4b9e-bbbe-a03e5316a562/score/up", httpCode: 401, message: "NotAuthorized: Your session is outdated. Please refresh or sync. [etc]

**Comment by paglias** [comment]
@TheHollidayInn I think the modal is appearing twice because the code to show it exists twice, here https://github.com/HabitRPG/habitica/blob/acf4b4da63e8c6666c589f0534f624d27116dfb8/website/client/components/notifications.vue#L216-L221 and https://github.com/HabitRPG/habitica/blob/acf4b4da63e8c6666c589f0534f624d27116dfb8/website/client/components/notifications.vue#L320-L323, can we just remove one of the two?

**Comment by TheHollidayInn** [comment]
Eh, we can't just remove one because one is reactive and one is a check after notifications are deferred for RYA. The section needs to be refsctored but it sounds like the death modal isn't updating the store or the user sync is off.

I can look into it more once your notifications PR is done because that should be cleaning up this section a lot.

**Comment by paglias** [comment]
Yeah but they all check user.stats.hp which is a reactive property so if this is called https://github.com/HabitRPG/habitica/blob/acf4b4da63e8c6666c589f0534f624d27116dfb8/website/client/components/notifications.vue#L320-L323 also this will be called https://github.com/HabitRPG/habitica/blob/acf4b4da63e8c6666c589f0534f624d27116dfb8/website/client/components/notifications.vue#L216-L221 because if user.stats.hp is < 0 the watcher on userHp gets called too

**Comment by TheHollidayInn** [comment]
checkUserAchievements is a method and is not called reactively. That is called manually after RYA is finished. But, either way, I'll test locally to see if I can recreate the issue because the second check should work. I added the first section because in some cases it wasn't, but that was a few months ago, so I'll revaluate.

**Comment by TheHollidayInn** [comment]
```
osberend
UserID: bc630499-2941-4c1b-85ed-41fed2e8959e
_Every time_ I die, whether on mobile or on a desktop browser, after the "You ran out of health!" message appears, I click okay, and the death process resolves, the "You ran out of health!" message immediately appears _again_, but clicking "okay" triggers the "can't revive if not dead" message. Usually just clicking outside the modal to dismiss it will fix the problem, but occasionally it will continue popping up at random times, despite my having non-zero health.
```

**Comment by lemoness** [comment]
Can I mark this as "help wanted" or is it still "needs investigation"?

**Comment by Alys** [comment]
This was probably from the site-wide bug of modals reappearing, which we have reported at https://github.com/HabitRPG/habitica/issues/9986

**Comment by noelo-cohelo** [comment]
I have this problem now on Android app.

**Comment by paglias** [comment]
Hi @noelo-cohelo and thanks for the report, could you report it directly from the app? In the app, Menu>About>Report a Bug. thanks!

**Comment by noelo-cohelo** [comment]
Thank you for your response @paglias. Unfortunately, I already uninstalled the app (partially because this was really annoying). But I will remember that I can use this option if I'll ever use it again.

**Comment by paglias** [comment]
Sorry about that @noelo-cohelo , in case you change your mind we're happy to help :)

**Comment by noelo-cohelo** [comment]
I'm sorry, I have trouble doing that.

When I click "report a bug", this is the effect and I don't know what to do
next.

[image: Screenshot_20200102-230747.png]

Best regards
Ewa

**Comment by paglias** [comment]
@noelo-cohelo sorry but I can't see your screenshot, what should happen is that a pre-filled email should open up to be sent to the mobile devs
